File: web/src/features/quickCapture.ts
import type { QuickCapturePreferences, TaskPriority } from "@/types";

export type QuickCaptureKind = "task" | "event";

export interface QuickCaptureRecognition {
  readonly id: string;
  readonly label: string;
  readonly removable: boolean;
}

export interface QuickCaptureResult {
  readonly kind: QuickCaptureKind;
  readonly rawTitle: string;
  readonly parsedTitle: string;
  readonly date?: string;
  readonly time?: string;
  readonly allDay: boolean;
  readonly eventDurationMinutes: number;
  readonly taskPriority: TaskPriority;
  readonly recurrence?: { readonly frequency: "daily" | "weekly" | "monthly" | "yearly"; readonly interval: number; readonly rrule: string };
  readonly recognitions: readonly QuickCaptureRecognition[];
  readonly eventReady: boolean;
}

interface Span extends QuickCaptureRecognition {
  readonly start: number;
  readonly length: number;
}

const defaultAliases: Pick<QuickCapturePreferences, "taskAliases" | "eventAliases" | "highPriorityAliases" | "mediumPriorityAliases" | "lowPriorityAliases"> = {
  taskAliases: ["task"],
  eventAliases: ["event"],
  highPriorityAliases: ["p1"],
  mediumPriorityAliases: ["p2"],
  lowPriorityAliases: ["p3"]
};

function localDate(value: Date): string {
  return `${value.getFullYear()}-${String(value.getMonth() + 1).padStart(2, "0")}-${String(value.getDate()).padStart(2, "0")}`;
}

function addDays(value: Date, amount: number): Date {
  const result = new Date(value);
  result.setDate(result.getDate() + amount);
  return result;
}

function aliasMatch(text: string, aliases: readonly string[]): RegExpMatchArray | undefined {
  const safe = aliases.map((alias) => alias.trim()).filter(Boolean).map((alias) => alias.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
  if (safe.length === 0) return undefined;
  return text.match(new RegExp(`\\b(?:${safe.join("|")})\\b`, "i")) ?? undefined;
}

function spanFor(match: RegExpMatchArray, prefix: string, label: string, removable = true): Span {
  return { id: `${prefix}:${match.index ?? 0}:${match[0].length}`, label, removable, start: match.index ?? 0, length: match[0].length };
}

function overlaps(left: Span, right: Span): boolean {
  return left.start < right.start + right.length && right.start < left.start + left.length;
}

function useSpan(spans: Span[], disabled: ReadonlySet<string>, span: Span, recognitions: QuickCaptureRecognition[]): boolean {
  if (disabled.has(span.id) || spans.some((existing) => overlaps(existing, span))) return false;
  spans.push(span);
  recognitions.push({ id: span.id, label: span.label, removable: span.removable });
  return true;
}

function parseDate(text: string, now: Date): { readonly date: string; readonly span: Span } | undefined {
  const numeric = /\b(\d{4})-(\d{2})-(\d{2})\b/.exec(text);
  if (numeric) {
    const candidate = new Date(Number(numeric[1]), Number(numeric[2]) - 1, Number(numeric[3]));
    if (candidate.getFullYear() === Number(numeric[1]) && candidate.getMonth() === Number(numeric[2]) - 1 && candidate.getDate() === Number(numeric[3])) {
      return { date: localDate(candidate), span: spanFor(numeric, "date", localDate(candidate)) };
    }
  }
  const today = /\btoday\b/i.exec(text);
  if (today) return { date: localDate(now), span: spanFor(today, "date", localDate(now)) };
  const tomorrow = /\btomorrow\b/i.exec(text);
  if (tomorrow) return { date: localDate(addDays(now, 1)), span: spanFor(tomorrow, "date", localDate(addDays(now, 1))) };
  const nextWeekday = /\bnext\s+(monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b/i.exec(text);
  if (nextWeekday) {
    const names = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"];
    const target = names.indexOf(nextWeekday[1]!.toLowerCase());
    const days = ((target - now.getDay() + 6) % 7) + 1;
    const date = addDays(now, days);
    return { date: localDate(date), span: spanFor(nextWeekday, "date", localDate(date)) };
  }
  const relative = /\bin\s+(\d{1,3})\s+(days?|weeks?)\b/i.exec(text);
  if (relative) {
    const date = addDays(now, Number(relative[1]) * (relative[2]!.toLowerCase().startsWith("week") ? 7 : 1));
    return { date: localDate(date), span: spanFor(relative, "date", localDate(date)) };
  }
  const named = /\b(january|february|march|april|may|june|july|august|september|october|november|december)\s+(\d{1,2})(?:st|nd|rd|th)?(?:,?\s+(\d{4}))?\b/i.exec(text);
  if (named) {
    const months = ["january", "february", "march", "april", "may", "june", "july", "august", "september", "october", "november", "december"];
    let year = named[3] ? Number(named[3]) : now.getFullYear();
    let date = new Date(year, months.indexOf(named[1]!.toLowerCase()), Number(named[2]));
    if (!named[3] && date < new Date(now.getFullYear(), now.getMonth(), now.getDate())) {
      year += 1;
      date = new Date(year, months.indexOf(named[1]!.toLowerCase()), Number(named[2]));
    }
    if (date.getMonth() === months.indexOf(named[1]!.toLowerCase()) && date.getDate() === Number(named[2])) return { date: localDate(date), span: spanFor(named, "date", localDate(date)) };
  }
  return undefined;
}

function parseTime(text: string): { readonly time: string; readonly span: Span } | undefined {
  const twelveHour = /\b(?:at\s+)?(\d{1,2})(?::(\d{2}))?\s*(am|pm)\b/i.exec(text);
  if (twelveHour) {
    let hour = Number(twelveHour[1]);
    const minute = Number(twelveHour[2] ?? "0");
    if (hour >= 1 && hour <= 12 && minute <= 59) {
      hour %= 12;
      if (twelveHour[3]!.toLowerCase() === "pm") hour += 12;
      const time = `${String(hour).padStart(2, "0")}:${String(minute).padStart(2, "0")}`;
      return { time, span: spanFor(twelveHour, "time", time) };
    }
  }
  const twentyFourHour = /\b(?:at\s+)?([01]?\d|2[0-3]):([0-5]\d)\b/.exec(text);
  if (!twentyFourHour) return undefined;
  const time = `${String(Number(twentyFourHour[1])).padStart(2, "0")}:${twentyFourHour[2]}`;
  return { time, span: spanFor(twentyFourHour, "time", time) };
}

function removeSpans(text: string, spans: readonly Span[]): string {
  let result = text;
  for (const span of [...spans].filter((item) => item.removable).sort((left, right) => right.start - left.start)) result = `${result.slice(0, span.start)}${result.slice(span.start + span.length)}`;
  return result.replace(/\s+/g, " ").trim();
}

export function parseQuickCapture(
  text: string,
  requestedKind: QuickCaptureKind,
  preferences: Pick<QuickCapturePreferences, "defaultEventDurationMinutes" | "removeRecognizedText" | "taskAliases" | "eventAliases" | "highPriorityAliases" | "mediumPriorityAliases" | "lowPriorityAliases"> = { defaultEventDurationMinutes: 30, removeRecognizedText: true, ...defaultAliases },
  disabledRecognitionIds: readonly string[] = [],
  now = new Date()
): QuickCaptureResult {
  const aliases = { ...defaultAliases, ...preferences };
  const disabled = new Set(disabledRecognitionIds);
  const spans: Span[] = [];
  const recognitions: QuickCaptureRecognition[] = [];
  let kind = requestedKind;
  const taskAlias = aliasMatch(text, aliases.taskAliases);
  const eventAlias = aliasMatch(text, aliases.eventAliases);
  const firstAlias = !taskAlias || eventAlias && (eventAlias.index ?? 0) < (taskAlias.index ?? 0)
    ? eventAlias && { match: eventAlias, kind: "event" as const, label: "Event" }
    : { match: taskAlias, kind: "task" as const, label: "Task" };
  if (firstAlias?.match) {
    const span = spanFor(firstAlias.match, "type", firstAlias.label);
    if (useSpan(spans, disabled, span, recognitions)) kind = firstAlias.kind;
  }
  let taskPriority: TaskPriority = "none";
  if (kind === "task") {
    for (const candidate of [
      { aliases: aliases.highPriorityAliases, priority: "high" as const, label: "High priority" },
      { aliases: aliases.mediumPriorityAliases, priority: "medium" as const, label: "Medium priority" },
      { aliases: aliases.lowPriorityAliases, priority: "low" as const, label: "Low priority" }
    ]) {
      const match = aliasMatch(text, candidate.aliases);
      if (match && useSpan(spans, disabled, spanFor(match, "priority", candidate.label), recognitions)) {
        taskPriority = candidate.priority;
        break;
      }
    }
  }
  let recurrence: QuickCaptureResult["recurrence"];
  const recurrenceMatch = /\bevery(?:\s+(\d{1,3}))?\s+(day|week|month|year)s?\b/i.exec(text);
  if (recurrenceMatch) {
    const interval = Number(recurrenceMatch[1] ?? "1");
    const frequency = `${recurrenceMatch[2]!.toLowerCase()}ly`.replace("dayly", "daily") as NonNullable<QuickCaptureResult["recurrence"]>["frequency"];
    const label = interval === 1 ? `Repeats every ${recurrenceMatch[2]!.toLowerCase()}` : `Repeats every ${interval} ${recurrenceMatch[2]!.toLowerCase()}s`;
    if (useSpan(spans, disabled, spanFor(recurrenceMatch, "recurrence", label), recognitions)) recurrence = { frequency, interval, rrule: `RRULE:FREQ=${frequency.toUpperCase()};INTERVAL=${interval}` };
  }
  let eventDurationMinutes = Math.max(1, Math.min(1_440, preferences.defaultEventDurationMinutes || 30));
  if (kind === "event") {
    const duration = /\bfor\s+(\d{1,4})\s*(m|min|mins|minute|minutes|h|hr|hrs|hour|hours)\b/i.exec(text);
    if (duration) {
      const minutes = Number(duration[1]) * (duration[2]!.toLowerCase().startsWith("h") ? 60 : 1);
      if (minutes >= 1 && minutes <= 1_440 && useSpan(spans, disabled, spanFor(duration, "duration", `${minutes} minutes`), recognitions)) eventDurationMinutes = minutes;
    }
  }
  const parsedDate = parseDate(text, now);
  let date = parsedDate?.date;
  if (parsedDate) useSpan(spans, disabled, parsedDate.span, recognitions);
  const parsedTime = parseTime(text);
  const time = parsedTime?.time;
  if (parsedTime && kind === "event") useSpan(spans, disabled, parsedTime.span, recognitions);
  if (parsedTime && kind === "task" && !disabled.has(parsedTime.span.id)) recognitions.push({ ...parsedTime.span, label: `${parsedTime.time} remains in task title`, removable: false });
  if (kind === "task" && recurrence && !date) date = localDate(now);
  if (kind === "event" && time && !date) {
    const [hour, minute] = time.split(":").map(Number);
    const todayAtTime = new Date(now.getFullYear(), now.getMonth(), now.getDate(), hour, minute);
    date = localDate(todayAtTime <= now ? addDays(now, 1) : now);
  }
  return {
    kind,
    rawTitle: text.trim(),
    parsedTitle: preferences.removeRecognizedText === false ? text.trim() : removeSpans(text, spans),
    date,
    time,
    allDay: kind === "event" && Boolean(date) && !time,
    eventDurationMinutes,
    taskPriority,
    recurrence,
    recognitions,
    eventReady: kind === "task" || Boolean(date)
  };
}

File: src/hcb/quick_capture.py
"""Natural-language quick capture parser.

The matching, span, and date/time rules follow the web TypeScript parser. Dates
and times are computed from a timezone-naive local datetime, using wall-clock
fields the same way JavaScript ``Date`` local getters do.
"""

from __future__ import annotations

import re
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Literal

from .models import CapturePreferences as QuickCapturePreferences

QuickCaptureKind = Literal["task", "event"]
TaskPriority = Literal["none", "low", "medium", "high"]
RecurrenceFrequency = Literal["daily", "weekly", "monthly", "yearly"]

_REGEX_FLAGS = re.IGNORECASE | re.ASCII
_ALIAS_ESCAPE = re.compile(r"[.*+?^${}()|[\]\\]")
_ISO_DATE = re.compile(r"\b(\d{4})-(\d{2})-(\d{2})\b", _REGEX_FLAGS)
_TODAY = re.compile(r"\btoday\b", _REGEX_FLAGS)
_TOMORROW = re.compile(r"\btomorrow\b", _REGEX_FLAGS)
_NEXT_WEEKDAY = re.compile(
    r"\bnext\s+(monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b",
    _REGEX_FLAGS,
)
_RELATIVE = re.compile(r"\bin\s+(\d{1,3})\s+(days?|weeks?)\b", _REGEX_FLAGS)
_NAMED_DATE = re.compile(
    r"\b(january|february|march|april|may|june|july|august|september|october|november|december)"
    r"\s+(\d{1,2})(?:st|nd|rd|th)?(?:,?\s+(\d{4}))?\b",
    _REGEX_FLAGS,
)
_TWELVE_HOUR = re.compile(r"\b(?:at\s+)?(\d{1,2})(?::(\d{2}))?\s*(am|pm)\b", _REGEX_FLAGS)
_TWENTY_FOUR_HOUR = re.compile(r"\b(?:at\s+)?([01]?\d|2[0-3]):([0-5]\d)\b", _REGEX_FLAGS)
_RECURRENCE = re.compile(r"\bevery(?:\s+(\d{1,3}))?\s+(day|week|month|year)s?\b", _REGEX_FLAGS)
_DURATION = re.compile(
    r"\bfor\s+(\d{1,4})\s*(m|min|mins|minute|minutes|h|hr|hrs|hour|hours)\b",
    _REGEX_FLAGS,
)
_WEEKDAYS = ("sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday")
_MONTHS = (
    "january",
    "february",
    "march",
    "april",
    "may",
    "june",
    "july",
    "august",
    "september",
    "october",
    "november",
    "december",
)
_FREQUENCIES: dict[str, RecurrenceFrequency] = {
    "day": "daily",
    "week": "weekly",
    "month": "monthly",
    "year": "yearly",
}


@dataclass(frozen=True, slots=True)
class QuickCaptureRecognition:
    id: str
    label: str
    removable: bool


@dataclass(frozen=True, slots=True)
class QuickCaptureRecurrence:
    frequency: RecurrenceFrequency
    interval: int
    rrule: str


@dataclass(frozen=True, slots=True)
class QuickCaptureResult:
    kind: QuickCaptureKind
    raw_title: str
    parsed_title: str
    all_day: bool
    event_duration_minutes: int
    task_priority: TaskPriority
    recognitions: tuple[QuickCaptureRecognition, ...]
    event_ready: bool
    date: str | None = None
    time: str | None = None
    recurrence: QuickCaptureRecurrence | None = None


@dataclass(frozen=True, slots=True)
class _Span:
    id: str
    label: str
    removable: bool
    start: int
    length: int

    def as_recognition(self) -> QuickCaptureRecognition:
        return QuickCaptureRecognition(id=self.id, label=self.label, removable=self.removable)


def _local_date(value: datetime) -> str:
    return f"{value.year:04d}-{value.month:02d}-{value.day:02d}"


def _add_days(value: datetime, amount: int) -> datetime:
    return value + timedelta(days=amount)


def _js_weekday(value: datetime) -> int:
    return (value.weekday() + 1) % 7


def _escape_alias(alias: str) -> str:
    return _ALIAS_ESCAPE.sub(lambda match: f"\\{match.group(0)}", alias)


def _alias_match(text: str, aliases: Sequence[str]) -> re.Match[str] | None:
    safe = [_escape_alias(alias.strip()) for alias in aliases if alias.strip()]
    if not safe:
        return None
    return re.search(rf"\b(?:{'|'.join(safe)})\b", text, _REGEX_FLAGS)


def _span_for(match: re.Match[str], prefix: str, label: str, removable: bool = True) -> _Span:
    start = match.start()
    length = len(match.group(0))
    return _Span(
        id=f"{prefix}:{start}:{length}",
        label=label,
        removable=removable,
        start=start,
        length=length,
    )


def _overlaps(left: _Span, right: _Span) -> bool:
    return left.start < right.start + right.length and right.start < left.start + left.length


def _use_span(
    spans: list[_Span],
    disabled: set[str],
    span: _Span,
    recognitions: list[QuickCaptureRecognition],
) -> bool:
    if span.id in disabled or any(_overlaps(existing, span) for existing in spans):
        return False
    spans.append(span)
    recognitions.append(span.as_recognition())
    return True


def _try_datetime(
    year: int, month: int, day: int, hour: int = 0, minute: int = 0
) -> datetime | None:
    try:
        return datetime(year, month, day, hour, minute)
    except ValueError:
        return None


def _parse_date(text: str, now: datetime) -> tuple[str, _Span] | None:
    numeric = _ISO_DATE.search(text)
    if numeric:
        year, month, day = (int(numeric.group(1)), int(numeric.group(2)), int(numeric.group(3)))
        candidate = _try_datetime(year, month, day)
        if candidate is not None:
            return _local_date(candidate), _span_for(numeric, "date", _local_date(candidate))

    today = _TODAY.search(text)
    if today:
        return _local_date(now), _span_for(today, "date", _local_date(now))

    tomorrow = _TOMORROW.search(text)
    if tomorrow:
        date = _add_days(now, 1)
        return _local_date(date), _span_for(tomorrow, "date", _local_date(date))

    next_weekday = _NEXT_WEEKDAY.search(text)
    if next_weekday:
        target = _WEEKDAYS.index(next_weekday.group(1).lower())
        days = ((target - _js_weekday(now) + 6) % 7) + 1
        date = _add_days(now, days)
        return _local_date(date), _span_for(next_weekday, "date", _local_date(date))

    relative = _RELATIVE.search(text)
    if relative:
        unit = relative.group(2).lower()
        date = _add_days(now, int(relative.group(1)) * (7 if unit.startswith("week") else 1))
        return _local_date(date), _span_for(relative, "date", _local_date(date))

    named = _NAMED_DATE.search(text)
    if named:
        month = _MONTHS.index(named.group(1).lower()) + 1
        day = int(named.group(2))
        year = int(named.group(3)) if named.group(3) else now.year
        named_date = _try_datetime(year, month, day)
        if named_date is None:
            return None
        if named.group(3) is None and named_date < datetime(now.year, now.month, now.day):
            rolled = _try_datetime(year + 1, month, day)
            if rolled is None:
                return None
            named_date = rolled
        return _local_date(named_date), _span_for(named, "date", _local_date(named_date))
    return None


def _parse_time(text: str) -> tuple[str, _Span] | None:
    twelve_hour = _TWELVE_HOUR.search(text)
    if twelve_hour:
        hour = int(twelve_hour.group(1))
        minute = int(twelve_hour.group(2) or "0")
        if 1 <= hour <= 12 and minute <= 59:
            hour %= 12
            if twelve_hour.group(3).lower() == "pm":
                hour += 12
            time = f"{hour:02d}:{minute:02d}"
            return time, _span_for(twelve_hour, "time", time)

    twenty_four_hour = _TWENTY_FOUR_HOUR.search(text)
    if twenty_four_hour is None:
        return None
    time = f"{int(twenty_four_hour.group(1)):02d}:{twenty_four_hour.group(2)}"
    return time, _span_for(twenty_four_hour, "time", time)


def _remove_spans(text: str, spans: Sequence[_Span]) -> str:
    result = text
    removable = sorted(
        (span for span in spans if span.removable),
        key=lambda span: span.start,
        reverse=True,
    )
    for span in removable:
        result = f"{result[: span.start]}{result[span.start + span.length :]}"
    return re.sub(r"\s+", " ", result).strip()


def parse_quick_capture(
    text: str,
    requested_kind: QuickCaptureKind,
    preferences: QuickCapturePreferences | None = None,
    disabled_recognition_ids: Sequence[str] = (),
    now: datetime | None = None,
) -> QuickCaptureResult:
    preferences = preferences or QuickCapturePreferences()
    current = now if now is not None else datetime.now()
    disabled = set(disabled_recognition_ids)
    spans: list[_Span] = []
    recognitions: list[QuickCaptureRecognition] = []
    kind: QuickCaptureKind = requested_kind

    task_alias = _alias_match(text, preferences.task_aliases)
    event_alias = _alias_match(text, preferences.event_aliases)
    first_alias: tuple[QuickCaptureKind, re.Match[str], str] | None
    if task_alias is None or (event_alias is not None and event_alias.start() < task_alias.start()):
        first_alias = ("event", event_alias, "Event") if event_alias is not None else None
    else:
        first_alias = ("task", task_alias, "Task")
    if first_alias is not None:
        match_kind, match, label = first_alias
        span = _span_for(match, "type", label)
        if _use_span(spans, disabled, span, recognitions):
            kind = match_kind

    task_priority: TaskPriority = "none"
    if kind == "task":
        priorities: tuple[tuple[tuple[str, ...], TaskPriority, str], ...] = (
            (preferences.high_priority_aliases, "high", "High priority"),
            (preferences.medium_priority_aliases, "medium", "Medium priority"),
            (preferences.low_priority_aliases, "low", "Low priority"),
        )
        for aliases, priority, label in priorities:
            priority_match = _alias_match(text, aliases)
            if priority_match and _use_span(
                spans, disabled, _span_for(priority_match, "priority", label), recognitions
            ):
                task_priority = priority
                break

    recurrence: QuickCaptureRecurrence | None = None
    recurrence_match = _RECURRENCE.search(text)
    if recurrence_match:
        interval = int(recurrence_match.group(1) or "1")
        unit = recurrence_match.group(2).lower()
        frequency = _FREQUENCIES[unit]
        label = f"Repeats every {unit}" if interval == 1 else f"Repeats every {interval} {unit}s"
        if _use_span(
            spans, disabled, _span_for(recurrence_match, "recurrence", label), recognitions
        ):
            recurrence = QuickCaptureRecurrence(
                frequency=frequency,
                interval=interval,
                rrule=f"RRULE:FREQ={frequency.upper()};INTERVAL={interval}",
            )

    event_duration_minutes = max(1, min(1_440, preferences.default_event_duration_minutes or 30))
    if kind == "event":
        duration = _DURATION.search(text)
        if duration:
            unit = duration.group(2).lower()
            minutes = int(duration.group(1)) * (60 if unit.startswith("h") else 1)
            if 1 <= minutes <= 1_440 and _use_span(
                spans,
                disabled,
                _span_for(duration, "duration", f"{minutes} minutes"),
                recognitions,
            ):
                event_duration_minutes = minutes

    parsed_date = _parse_date(text, current)
    date = parsed_date[0] if parsed_date else None
    if parsed_date:
        _use_span(spans, disabled, parsed_date[1], recognitions)

    parsed_time = _parse_time(text)
    time = parsed_time[0] if parsed_time else None
    if parsed_time and kind == "event":
        _use_span(spans, disabled, parsed_time[1], recognitions)
    if parsed_time and kind == "task" and parsed_time[1].id not in disabled:
        recognitions.append(
            QuickCaptureRecognition(
                id=parsed_time[1].id,
                label=f"{parsed_time[0]} remains in task title",
                removable=False,
            )
        )
    if kind == "task" and recurrence is not None and date is None:
        date = _local_date(current)
    if kind == "event" and time is not None and date is None:
        hour, minute = (int(part) for part in time.split(":"))
        today_at_time = datetime(current.year, current.month, current.day, hour, minute)
        date = _local_date(_add_days(current, 1) if today_at_time <= current else current)

    parsed_title = (
        text.strip() if preferences.remove_recognized_text is False else _remove_spans(text, spans)
    )
    return QuickCaptureResult(
        kind=kind,
        raw_title=text.strip(),
        parsed_title=parsed_title,
        date=date,
        time=time,
        all_day=kind == "event" and bool(date) and not time,
        event_duration_minutes=event_duration_minutes,
        task_priority=task_priority,
        recurrence=recurrence,
        recognitions=tuple(recognitions),
        event_ready=kind == "task" or bool(date),
    )
